### ISSUE TITLE
Selectively export or copy portion of audio file to separate file

### ISSUE THREAD

**Issue by psideas** [naudio/NAudio#255]
Hi,

I have an audio file that may be an hour or longer and I want to export multiple ranges (regions of time within the file) to a separate file. Other times, I want each range to export as its own file (ie. 3 ranges becomes 3 files). I am specifying ranges in milliseconds elapsed since the beginning of the file.

The original file from which I am selectively exporting is WAV (either 44.1K or 48K). The destination file would inherit the format, sample rate, and channels from the original file.

Any help would be greatly appreciated.

Thanks

**Comment by markheath** [comment]
A few ways you can do this. Here's a low-level approach keeping the format exactly the same and will allow millisecond accuracy


```
var inputFile = @"D:\example.wav";
var outputFile = @"D:\trimmed.wav";
var startTime = TimeSpan.FromSeconds(4);
var duration = TimeSpan.FromSeconds(10);
using(var reader = new WaveFileReader(inputFile))
{
	reader.CurrentTime = startTime;
	var bytesToRead = (int)(reader.WaveFormat.AverageBytesPerSecond * duration.TotalSeconds);
	bytesToRead -= (bytesToRead % reader.WaveFormat.BlockAlign);
	var buffer = new byte[reader.WaveFormat.AverageBytesPerSecond * 4];
	using (var writer = new WaveFileWriter(outputFile, reader.WaveFormat))
	{
		var copied = 0;
		while (copied < bytesToRead)
		{
			var wanted = Math.Min(buffer.Length, bytesToRead - copied);
			var read = reader.Read(buffer,0,wanted);
			if (read == 0) break;
			writer.Write(buffer,0,read);
			copied+=read;
		}
	}
}

**Comment by psideas** [comment]
This worked perfectly. Thanks.

**Comment by psideas** [comment]
Hi Mark,

I have used this code snippet in the past with success but I am now working on Windows service (still in C#) and I am encountering an issue. I try to selectively export a part of a file and get zero-byte files.

![image](https://user-images.githubusercontent.com/30737800/53674779-5c9e9580-3c5e-11e9-9175-206e0c3bfd74.png)

I notice that there is a break statement that is terminating the while loop immediately. Any thoughts on that?

For your reference, I have changed your var statements a bit to reflect how I am trying to use it:

>var inputFile = Properties.Settings.Default.ScanDir + "\\" + CurrentAudioFileSpec;
>var outputFile = Properties.Settings.Default.DownloadDir + "\\" + CurrentReference + "_" + EventId + ".wav";
>var startTime = TimeSpan.FromMilliseconds(Convert.ToDouble(StartOffset));
>var duration = TimeSpan.FromMilliseconds(Convert.ToDouble(EndOffset) - Convert.ToDouble(StartOffset));

Is there perhaps some other unit I need to change from seconds to milliseconds? Any thoughts would be greatly appreciated.

Thanks,
Paul
